### PR TITLE
Add estimated and actual block cost units metrics

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem
Want to identify the delta between leader estimated block units and actual units after execution. 
 
#### Summary of Changes
- report estimated execution units packed into block, and actual units and time after execution.
- report roll-up transaction costs per block, give us info such as "how many write lock units in this block"

Fixes #

### CODE REVIEW

File: core/src/qos_service.rs [hunk 1]
```diff
@@ -128,6 +143,16 @@ impl QosService {
...
+                    self.report_sender
+                        .send(QosMetrics::TransactionCost {
+                            slot: bank.slot(),
+                            tx_signature: *tx.signature(),
+                            signature_cost: cost.signature_cost,
+                            write_lock_cost: cost.write_lock_cost,
+                            data_bytes_cost: cost.data_bytes_cost,
+                            execution_cost: cost.execution_cost,
+                        } )
+                        .unwrap_or_else(|err| warn!("qos service report tx cost details failed: {:?}", err));
```
Comment (taozhu-chicago) [inline core/src/qos_service.rs:155]: not super comfortable of sending every transaction cost details to influx, so just sending the needed CUs for now.

File: core/src/qos_service.rs [hunk 2]
```diff
@@ -194,14 +219,56 @@ impl QosService {
...
+                        slot,
+                        tx_signature,
+                        signature_cost,
+                        write_lock_cost,
+                        data_bytes_cost,
+                        execution_cost,
+                    } => {
+                        // report transaction cost details per slot|signature
+                        datapoint_info!(
+                            "qos-service-transaction-costs",
```
Comment (sakridge) [inline core/src/qos_service.rs:261]: we can't send every transaction cost to the metrics

**Comment by sakridge** [comment]
We'll have gaps in the data if only the leader sends the metrics. Not everyone has metrics enabled.

**Comment by taozhu-chicago** [comment]
> We'll have gaps in the data if only the leader sends the metrics. Not everyone has metrics enabled.

Yeah, it's just a lot of data for metrics. Could print transaction costs in ledger-tool, but have to plot outside of grafana, making link to instruction data, all the goodies Trent is adding, much more difficult.

**Comment by sakridge** [comment]
> 

Can we just add it per-block?

**Comment by sakridge** [comment]
I think 1 datapoint per transaction is still excessive even for 1 node, but something like this submission could be enabled optionally in the replay path and then we have one of our validators pass the option to upload the data.

**Comment by taozhu-chicago** [comment]
> 



> I think 1 datapoint per transaction is still excessive even for 1 node, but something like this submission could be enabled optionally in the replay path and then we have one of our validators pass the option to upload the data.

Agree, data point per transaction is too much data. I ran cluster test, the log files were filled with "transaction cost"s. 
What do you mean by per-block? Logging all transactions once per block? That'd still bust db, isn't it?

**Comment by sakridge** [comment]
> > I think 1 datapoint per transaction is still excessive even for 1 node, but something like this submission could be enabled optionally in the replay path and then we have one of our validators pass the option to upload the data.
> 
> Agree, data point per transaction is too much data. I ran cluster test, the log files were filled with "transaction cost"s. What do you mean by per-block? Logging all transactions once per block? That'd still bust db, isn't it? Like the idea of doing it in replay with option to turn it on/off. Think this is probably is the best options, even it does not paint the exact picture leaders sees, it's gonna be pretty close. I have removed the per transaction reporting, adding similar function to ledger-tool. Left the estimated cost/actual cost so we will know how off leader's estimates are

For per-block I mean to roll-up all the data for a block and then submit that. Like total CU used per block, or total write locks per block, etc. Then you can take the average used per transaction or instruction or whatever.

**Comment by t-nelson** [comment]
> I think 1 datapoint per transaction is still excessive even for 1 node, but something like this submission could be enabled optionally in the replay path and then we have one of our validators pass the option to upload the data.

`datapoint_trace!()`

**Comment by taozhu-chicago** [comment]
> > I think 1 datapoint per transaction is still excessive even for 1 node, but something like this submission could be enabled optionally in the replay path and then we have one of our validators pass the option to upload the data.
> 
> `datapoint_trace!()`

Brilliant! this would work the same as option for validator. The difference is, with option for replay to upload, we have control of when to collect data; With datapoint_trace on banking stage, we will only have data when our node is scheduled as leader. For troubleshooting/analyzing, it sounds better to do so with optional replay reporting?

**Comment by sakridge** [comment]
> > > I think 1 datapoint per transaction is still excessive even for 1 node, but something like this submission could be enabled optionally in the replay path and then we have one of our validators pass the option to upload the data.
> > 
> > 
> > `datapoint_trace!()`
> 
> Brilliant! this would work the same as option for validator. The difference is, with option for replay to upload, we have control of when to collect data; With datapoint_trace on banking stage, we will only have data when our node is scheduled as leader. For troubleshooting/analyzing, it sounds better to do so with optional replay reporting?

yea, I think so.

**Comment by taozhu-chicago** [comment]
> > > > I think 1 datapoint per transaction is still excessive even for 1 node, but something like this submission could be enabled optionally in the replay path and then we have one of our validators pass the option to upload the data.
> > > 
> > > 
> > > `datapoint_trace!()`
> > 
> > 
> > Brilliant! this would work the same as option for validator. The difference is, with option for replay to upload, we have control of when to collect data; With datapoint_trace on banking stage, we will only have data when our node is scheduled as leader. For troubleshooting/analyzing, it sounds better to do so with optional replay reporting?
> 
> yea, I think so.

Updated the PR description to just focus on adding estimated/actual units per block (this allows us to know how off leader estimations are), and rolled up transaction cost per block (eg. how many signature units in this block, etc).

File: core/src/banking_stage.rs [hunk 3]
```diff
@@ -1010,6 +1022,51 @@ impl BankingStage {
         (result, retryable_txs)
     }
 
+    // rollup transaction cost details, eg signature_cost, write_lock_cost, data_bytes_cost and
+    // execution_cost from the batch of transactions selected for block.
+    fn accumulate_batched_transaction_costs<'a>(
+        transactions_costs: impl Iterator<Item = &'a TransactionCost>,
+        transaction_results: impl Iterator<Item = &'a transaction::Result<()>>,
+    ) -> (u64, u64, u64, u64) {
```
Comment (t-nelson) [inline core/src/banking_stage.rs:1030]: Struct please!

File: core/src/banking_stage.rs [hunk 4]
```diff
@@ -1010,6 +1022,51 @@ impl BankingStage {
...
+        let ((signature_costs, write_lock_costs), (data_bytes_costs, execution_costs)): (
+            (Vec<_>, Vec<_>),
+            (Vec<_>, Vec<_>),
```
Comment (t-nelson) [inline core/src/banking_stage.rs:1033]: What's the value of these inner tuples?

File: core/src/qos_service.rs [hunk 5]
```diff
@@ -260,6 +301,24 @@ struct QosServiceMetrics {
...
+    estimated_data_bytes_units: AtomicU64,
+
+    // accumulated estimated program Compute Units to be packed into block
+    estimated_execute_units: AtomicU64,
+
+    // accumulated actual program Compute Units that have been packed into block
+    actual_execute_units: AtomicU64,
+
+    // accumulated actual program execute micro-sec that have been packed into block
+    actual_execute_time: AtomicU64,
```
Comment (t-nelson) [inline core/src/qos_service.rs:321]:
```suggestion
    actual_execute_time_us: AtomicU64,
```

File: core/src/qos_service.rs [hunk 6]
```diff
@@ -260,6 +301,24 @@ struct QosServiceMetrics {
...
+    estimated_write_lock_units: AtomicU64,
+
+    // accumulated estimated instructino data Compute Units to be packed into block
+    estimated_data_bytes_units: AtomicU64,
+
+    // accumulated estimated program Compute Units to be packed into block
+    estimated_execute_units: AtomicU64,
+
+    // accumulated actual program Compute Units that have been packed into block
+    actual_execute_units: AtomicU64,
```
Comment (t-nelson) [inline core/src/qos_service.rs:318]: Can we start calling these `_cu` instead of ambiguous `_units`?

File: core/src/qos_service.rs [hunk 7]
```diff
@@ -339,6 +398,36 @@ impl QosServiceMetrics {
...
+                    i64
+                ),
+                (
+                    "actual_execute_units",
+                    self.actual_execute_units.swap(0, Ordering::Relaxed) as i64,
+                    i64
+                ),
+                (
+                    "actual_execute_time",
+                    self.actual_execute_time.swap(0, Ordering::Relaxed) as i64,
```
Comment (t-nelson) [inline core/src/qos_service.rs:428]:
```suggestion
                    "actual_execute_time_us",
                    self.actual_execute_time_us.swap(0, Ordering::Relaxed) as i64,
```

File: core/src/qos_service.rs [hunk 8]
```diff
@@ -194,14 +198,51 @@ impl QosService {
...
+
+    pub fn accumulate_actual_execute_units(&self, units: u64) {
+        self.metrics
+            .actual_execute_units
+            .fetch_add(units, Ordering::Relaxed);
+    }
+
+    pub fn accumulate_actual_execute_time(&self, micro_sec: u64) {
+        self.metrics
+            .actual_execute_time
```
Comment (t-nelson) [inline core/src/qos_service.rs:230]:
```suggestion
            .actual_execute_time_us
```

**Review by t-nelson — DISMISSED** [review]
lgtm, just naming nits

File: core/src/banking_stage.rs (hunk 3 above)
Comment (taozhu-chicago) [reply core/src/banking_stage.rs]: changed

File: core/src/qos_service.rs (hunk 6 above)
Comment (taozhu-chicago) [reply core/src/qos_service.rs]: Yeah, time of `cu` has arrived 😄

File: core/src/qos_service.rs (hunk 8 above)
Comment (taozhu-chicago) [reply core/src/qos_service.rs]: Added _us to all places.

**Pull request merged**